### PR TITLE
Optimize ConvTransposed2D with mkldnn float32 and bfloat16 on CPU

### PR DESCRIPTION
**Pull request by mingfeima**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #92530

this PR optimized `ConvTranspose2d` with oneDNN and add channels last support for it. Also the fallback path `slow_conv_transpose2d` also have channels last support. So the memory format propagation behavior would stay the same with or without oneDNN.

Replacement of https://github.com/pytorch/pytorch/pull/77060, https://github.com/pytorch/pytorch/pull/70897 and https://github.com/pytorch/pytorch/pull/74023 which enables oneDNN for `ConvTranspose2d` and `ConvTranspose3d`

The following results collects on Skylake Xeon 8180, dual sockets, 28 cores per socket.
### single core channels last

configs | forward before/ms | forward after/ms | ratio | backward   before/ms | backward after/ms | ratio
-- | -- | -- | -- | -- | -- | --
input size: (32, 32, 100, 100), weight size: (32, 32, 3, 3) | 181.36 | 91.16 | 1.99 | 531.38 | 124.08 | 4.28
input size:   (32, 16, 200, 200), weight size: (16, 16, 3, 3) | 324.35 | 153.50 | 2.11 | 973.16 | 185.97 | 5.23
input size:   (32, 128, 100, 100), weight size: (128, 128, 3, 3) | 1086.82 | 671.52 | 1.62 | 3008.94 | 1453.33 | 2.07

### single core channels first

configs | forward before/ms | forward after/ms | ratio | backward   before/ms | backward after/ms | ratio
-- | -- | -- | -- | -- | -- | --
input size: (32, 32, 100, 100), weight size: (32, 32, 3, 3) | 138.10 | 5.94 | 23.23 | 37.97 | 11.25 | 3.38
input size:   (32, 16, 200, 200), weight size: (16, 16, 3, 3) | 236.43 | 8.75 | 27.03 | 87.77 | 18.58 | 4.72
input size:   (32, 128, 100, 100), weight size: (128, 128, 3, 3) | 484.39 | 37.69 | 12.85 | 185.40 | 90.57 | 2.05

### single socket channels last

configs | forward before/ms | forward after/ms | ratio | backward   before/ms | backward after/ms | ratio
-- | -- | -- | -- | -- | -- | --
input size: (32, 32, 100, 100), weight size: (32, 32, 3, 3) | 138.10 | 5.94 | 23.23 | 37.97 | 11.25 | 3.38
input size:   (32, 16, 200, 200), weight size: (16, 16, 3, 3) | 236.43 | 8.75 | 27.03 | 87.77 | 18.58 | 4.72
input size:   (32, 128, 100, 100), weight size: (128, 128, 3, 3) | 484.39 | 37.69 | 12.85 | 185.40 | 90.57 | 2.0

### single socket channels first

configs | forward before/ms | forward after/ms | ratio | backward   before/ms | backward after/ms | ratio
-- | -- | -- | -- | -- | -- | --
input size: (32, 32, 100,   100), weight size: (32, 32, 3, 3) | 132.56 | 7.19 | 18.43 | 31.43 | 11.20 | 2.81
input size:   (32, 16, 200, 200), weight size: (16, 16, 3, 3) | 227.94 | 13.33 | 17.11 | 63.00 | 23.41 | 2.69
input size:   (32, 128, 100, 100), weight size: (128, 128, 3, 3) | 473.68 | 52.79 | 8.97 | 150.40 | 87.33 | 1.72

cc @jgong5 @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: aten/src/ATen/native/Convolution.cpp [hunk 1]
```diff
@@ -500,13 +500,18 @@ struct ConvParams {
     if (!at::globalContext().userEnabledMkldnn()) {
       return false;
     }
+    if (transposed && is_output_padding_big()) {
+      return false;
+    }
+    if (transposed && groups > 1 && at::symint::size<T>(input, 1) == groups) {
+      return false;
+    }
```
Comment (jgong5) [inline aten/src/ATen/native/Convolution.cpp:508]: Add comments on the reason why we want fallback here?

**Review by jgong5 — APPROVED** [review]
LGTM.

File: aten/src/ATen/native/Convolution.cpp (hunk 1 above)
Comment (mingfeima) [reply aten/src/ATen/native/Convolution.cpp:508]: bug in `ideep` that recognizes a 4d tensor with groups (and C=1) to be a channels_last_3d tensor. in oneDNN, g is treated as a separate dimension.

will have it fixed.

**Comment by mingfeima** [comment]
@ezyang need  a super user here, probably because this one updates `test/nn/test_convolution.py`.

File: aten/src/ATen/native/NaiveConvolutionTranspose2d.cpp [hunk 2]
```diff
@@ -291,98 +290,97 @@ void slow_conv_transpose2d_out_cpu_template(
...
+    });
+  });
+
+  if (bias.defined()) {
+    output.add_(bias_.reshape({-1, 1, 1}));
+  }
+
+  // Resize output
+  if (is_batch) {
+    output.resize_({n_output_plane, output_height, output_width});
```
Comment (ezyang) [inline aten/src/ATen/native/NaiveConvolutionTranspose2d.cpp:382]: It's not clear to me why you bothered optimizing the naive convolution transpose lol

**Review by ezyang — APPROVED** [review]
I did not review algorithm, just checking all framework'y bits were ok

File: aten/src/ATen/native/NaiveConvolutionTranspose2d.cpp (hunk 2 above)
Comment (mingfeima) [reply aten/src/ATen/native/NaiveConvolutionTranspose2d.cpp:382]: From the functional respective, we want to make sure a layer with Channels last support would have the same memory format propagation behavior `with` or `without` mkldnn. Just adding some fundamental optimization for the naive path, did not take too much time.

**Pull request closed without merging**